services/cannon/src/Cannon/RabbitMqConsumerApp.hs

```diff
@@ -3,22 +3,78 @@
 module Cannon.RabbitMqConsumerApp where
 
 import Cannon.App (rejectOnError)
+import Cannon.Dict qualified as D
+import Cannon.Options
 import Cannon.WS hiding (env)
-import Cassandra as C
+import Cassandra as C hiding (batch)
 import Control.Concurrent.Async
+import Control.Concurrent.Timeout
 import Control.Exception (Handler (..), bracket, catch, catches, throwIO, try)
+import Control.Lens hiding ((#))
 import Control.Monad.Codensity
-import Data.Aeson
+import Data.Aeson hiding (Key)
 import Data.Id
-import Imports
+import Data.List.Extra hiding (delete)
+import Data.Timeout (TimeoutUnit (..), (#))
+import Imports hiding (min, threadDelay)
 import Network.AMQP qualified as Q
 import Network.AMQP.Extended (withConnection)
 import Network.WebSockets
 import Network.WebSockets qualified as WS
 import System.Logger qualified as Log
+import UnliftIO.Async (pooledMapConcurrentlyN_)
 import Wire.API.Event.WebSocketProtocol
 import Wire.API.Notification
 
+drainRabbitQueues :: Env -> IO ()
+drainRabbitQueues e = do
+  conns <- D.toList e.rabbitConnections
+  numberOfConns <- fromIntegral <$> D.size e.rabbitConnections
+
+  let opts = e.drainOpts
+      maxNumberOfBatches = (opts ^. gracePeriodSeconds * 1000) `div` (opts ^. millisecondsBetweenBatches)
+      computedBatchSize = numberOfConns `div` maxNumberOfBatches
+      batchSize = max (opts ^. minBatchSize) computedBatchSize
+
+  logDraining e.logg numberOfConns batchSize (opts ^. minBatchSize) computedBatchSize maxNumberOfBatches
+
+  -- Sleeps for the grace period + 1 second. If the sleep completes, it means
+  -- that draining didn't finish, and we should log that.
+  timeoutAction <- async $ do
+    -- Allocate 1 second more than the grace period to allow for overhead of
+    -- spawning threads.
+    liftIO $ threadDelay $ ((opts ^. gracePeriodSeconds) # Second + 1 # Second)
+    logExpired e.logg (opts ^. gracePeriodSeconds)
+
+  for_ (chunksOf (fromIntegral batchSize) conns) $ \batch -> do
+    -- 16 was chosen with a roll of a fair dice.
+    void . async $ pooledMapConcurrentlyN_ 16 (uncurry (closeConn e.logg)) batch
+    liftIO $ threadDelay ((opts ^. millisecondsBetweenBatches) # MilliSecond)
+  cancel timeoutAction
+  Log.info e.logg $ Log.msg (Log.val "Draining complete")
+  where
+    closeConn :: Log.Logger -> Key -> Q.Connection -> IO ()
+    closeConn l key conn = do
+      Log.info l $
+        Log.msg (Log.val "closing rabbitmq connection")
+          . Log.field "key" (show key)
+      Q.closeConnection conn
+      void $ D.remove key e.rabbitConnections
```

Comment:
What happens if Cannon crashes in between these two?

Reply:
If it completely crashes, then there's no issue, it's not like the in-memory data will be preserved. Since this is bookkeeping for the graceful termination. If it starts graceful termination between these two steps, the worst case is trying to close a connection twice, which should be a no-op.